azurerm/resource_arm_virtual_network_gateway.go

```diff
@@ -0,0 +1,686 @@
+	// has finished, this workaround was suggested by Azure support to avoid conflicts
+	// when modifying/deleting the related subnet or network. Although the API indicated
+	// that the Virtual Network Gateway does not exist anymore, there is still a link
+	// to the Gateway Subnet it has been associated with. This causes an error when
+	// trying to delete the Gateway Subnet immediately after the Virtual Network Gateway
+	// has been deleted.
+
+	d.SetId("")
+	return nil
+}
+
+func virtualNetworkGatewayStateRefreshFunc(client *ArmClient, resourceGroupName string, virtualNetworkGateway string, withNotFound bool) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		res, err := client.vnetGatewayClient.Get(resourceGroupName, virtualNetworkGateway)
+		if err != nil {
+			if withNotFound && res.StatusCode == http.StatusNotFound {
+				return res, "NotFound", nil
+			}
+			return nil, "", fmt.Errorf("Error issuing read request in virtualNetworkGatewayStateRefreshFunc to Azure ARM for VirtualNetworkGateway '%s' (RG: '%s'): %s", virtualNetworkGateway, resourceGroupName, err)
+		}
+
+		return res, *res.VirtualNetworkGatewayPropertiesFormat.ProvisioningState, nil
+	}
+}
+
+func getArmVirtualNetworkGatewayProperties(d *schema.ResourceData) (*network.VirtualNetworkGatewayPropertiesFormat, error) {
+	gatewayType := network.VirtualNetworkGatewayType(d.Get("type").(string))
+	vpnType := network.VpnType(d.Get("vpn_type").(string))
+	enableBgp := d.Get("enable_bgp").(bool)
+	activeActive := d.Get("active_active").(bool)
+
+	props := &network.VirtualNetworkGatewayPropertiesFormat{
+		GatewayType:      gatewayType,
+		VpnType:          vpnType,
+		EnableBgp:        &enableBgp,
+		ActiveActive:     &activeActive,
+		Sku:              expandArmVirtualNetworkGatewaySku(d),
+		IPConfigurations: expandArmVirtualNetworkGatewayIPConfigurations(d),
+	}
+
+	if gatewayDefaultSiteID := d.Get("default_local_network_gateway_id").(string); gatewayDefaultSiteID != "" {
+		props.GatewayDefaultSite = &network.SubResource{
+			ID: &gatewayDefaultSiteID,
+		}
+	}
+
+	if _, ok := d.GetOk("vpn_client_configuration"); ok {
+		props.VpnClientConfiguration = expandArmVirtualNetworkGatewayVpnClientConfig(d)
+	}
+
+	if _, ok := d.GetOk("bgp_settings"); ok {
+		props.BgpSettings = expandArmVirtualNetworkGatewayBgpSettings(d)
+	}
+
+	// Sku validation for policy-based VPN gateways
+	if props.GatewayType == network.VirtualNetworkGatewayTypeVpn && props.VpnType == network.PolicyBased {
+		ok, err := evaluateSchemaValidateFunc(string(props.Sku.Name), "sku", validateArmVirtualNetworkGatewayPolicyBasedVpnSku())
+
+		if !ok {
+			return nil, err
+		}
+	}
+
+	// Sku validation for route-based VPN gateways
+	if props.GatewayType == network.VirtualNetworkGatewayTypeVpn && props.VpnType == network.RouteBased {
+		ok, err := evaluateSchemaValidateFunc(string(props.Sku.Name), "sku", validateArmVirtualNetworkGatewayRouteBasedVpnSku())
+
+		if !ok {
+			return nil, err
+		}
+	}
```

Comment:
I found https://github.com/hashicorp/terraform/issues/12021 and an older pull request which has been closed https://github.com/hashicorp/terraform/pull/4348. Reads like there is something blocking to implement resource level validation.